packages/inferno-mobx/src/observerWrap.ts

```diff
@@ -0,0 +1,116 @@
+import { InfernoNode, render, VNode, createComponentVNode, _HI as normalizeRoot } from 'inferno';
+import { Reaction } from 'mobx';
+import { throwError, warning } from 'inferno-shared';
+import { VNodeFlags } from 'inferno-vnode-flags';
+
+type Render = (properties?: any, context?: Record<string, unknown>) => InfernoNode | undefined | void;
+
+function callDispose({ dispose }: { readonly dispose: () => void }): void {
+  dispose();
+}
+
+interface InnerProperties {
+  readonly context: unknown
+  readonly dispose: () => void
+  readonly props: unknown
+  readonly self: VNode
+  readonly track: (f: () => void) => void
+}
+
+function innerVNode<T>(type: (p: InnerProperties) => T, properties: InnerProperties): VNode {
+  const ref = {
+    onComponentDidUpdate: callDispose,
+    onComponentWillUnmount: properties.dispose
+  };
+  if (process.env.NODE_ENV !== 'production') {
+    Object.freeze(properties);
+    Object.freeze(ref);
+  }
+  return createComponentVNode(VNodeFlags.ComponentFunction, type, properties, undefined, ref);
+}
+
+function makeProxy(target: VNode): { $V: InfernoNode } {
+  return {
+    get $V() {
+      return target.children
+    },
+    set $V(value) {
+      target.children = value
+    }
+  };
+}
+
+export function observerWrap<T extends Render>(base: T): typeof base {
+  if (process.env.NODE_ENV !== 'production') {
+    if (typeof base !== 'function') {
+      throwError(`observerWrap requires a function to wrap, got ${typeof base} instead`);
+    }
+    // @ts-ignore
+    if (base.prototype && base.prototype.render) {
+      throwError("observerWrap should not be applied to constructors.");
+    }
+    // @ts-ignore
+    if (base.isMobXInfernoObserver) {
+      warning("'observerWrap' was used on a component that already has 'observerWrap' applied. Please only apply once");
+    }
+  }
+  function tracked({ context, props, self, track }: InnerProperties): ReturnType<typeof base> {
+    let result;
+    let caught;
+    track(() => {
+      try {
+        result = base.call(self, props, context);
+      } catch(error) {
+        caught = error;
+      }
+    });
+    if (caught) {
+      throw caught;
+    }
+    return result;
+  }
+  function wrapper(this: VNode, props, context): VNode {
+    let onComponentDidUpdate;
+    let onComponentWillUpdate;
+    if (this.ref) {
+      const ref = this.ref;
+      if (ref.onComponentDidUpdate) {
+        onComponentDidUpdate = ref.onComponentDidUpdate.bind(ref, props, props);
```

Comment:
These props should be prev and next props is it really working if method is bound like this?

Reply:
When the MobX reaction re-renders the component, no new `props` will have been received. Instead some of the values will have changed.

If Inferno re-renders the component, the reaction is to be disposed of and replaced with a new one which will have the latest `props`.
Except when an `onComponentShouldUpdate` hook keeps that from happening, in which case the old and new `props` should be equivalent.